### ISSUE TITLE
resource viewer CRUD

### ISSUE THREAD

**Issue by aryam7** [aryam7/studycloud#83]
![20180420_221959](https://user-images.githubusercontent.com/23412689/43031870-33ebd5e2-8c70-11e8-8e31-4633a423d01a.jpg)

**Comment by aryam7** [comment]
DateTimes should be formatted as `April 1, 2018 1:00 AM` when returned in JSON.

**Comment by aryam7** [comment]
I've started work on this in the `resourceViewerCRUD` branch. See commit f1cfc63c0f9ddd37195d7d03fc0d425f677868fa.
@GiselleSerate and @ciennG, I may have changed the URL slightly from what we had in the image above. Instead of `/resource/`, I used `/resources/`, which is more consistent with the standard format of Laravel CRUD URLs. This means, for example, that the GET request to show resource 1 will now be `/resources/1` instead of `/resource/1`.

**Comment by GiselleSerate** [comment]
(Assigning myself so I remember to verify these changes against the code I wrote, unless Cienn gets on it first.)

**Comment by aryam7** [comment]
Ok. I've got a couple more changes to the routes, actually. Sorry about all of this.

We originally planned for `/resources/create` to be the route for submitting a request to store something in the database. But I just realized that we probably need `/resources/create` for showing the resource creation page. So we'll use a POST request to `/resources` as the route for submitting a request to create a new resource instead.
Also, there really wasn't any point in having the request for JSON be a POST request, so I made it a GET instead.

To summarize all of our routes in their final form:

HTTP Verb | URI | Action | Route Name
--- | --- | --- | ---
GET | `/resources/create` | show the resource creation page | resources.create
POST | `/resources` | create a new resource sent as JSON | resources.store
GET | `/resources/{id}` | show the page for this resource (and the editor if logged in as the author) | resources.show
GET | `/resources/data/{id}` | get the JSON representation of this resource | resources.json
PATCH (or PUT) | `resources/{id}` | alter a current resource according to the changes sent as JSON | resources.update
DELETE | `/resources/{id}` | request that this resource be deleted | resources.destroy

**Comment by aryam7** [comment]
I may add some more in the future, in which case I will update the table above.
Here are some that I'm thinking about. Let me know what you think.

HTTP Verb | URI | Action | Route Name
--- | --- | --- | ---
POST | `/resources/attach/{id}` | add this resource to a list of topics sent as JSON (overriding any conflicting topics that are currently attached) | resources.attach
POST | `/resources/detach/{id}` | remove this resource from the list of topics sent as JSON | resources.detach

**Comment by aryam7** [comment]
Ok. I've decided to make a couple more changes. These will only affect how the JSON representation of this resource is queried. Hopefully, this will also set a standard for how JSON data is retrieved from the server in the future.
1. Rather than use a POST request to get the JSON representation of this resource, we should be making a GET request, since the action performed by the request isn't changing anything in the database. ([source](https://restfulapi.net/http-methods/))
2. In the future, a straightforward way of identifying which URLs to get JSON data from would be to just append `/data` to them (as in `/resources/data` if you want JSON data related to resources). This is a standard already maintained by the topic tree (ie `tree/data` will give you JSON data related to the topic tree).
It might be best, then, to leave off the `{id}` so that `/resources/data` can be the entry point URL for any data related to resources (even data that isn't associated with a single resource). Instead, we can use the query `id` to specify that we are making a request for data related to a specific resource.

To summarize, this was what the request looked like before:

HTTP Verb | URI | Action | Route Name
--- | --- | --- | ---
POST | `/resources/data/{id}` | get the JSON representation of this resource | resources.json

And this is what it will look like now:

HTTP Verb | URI | Action | Route Name
--- | --- | --- | ---
GET | `/resources/data?id=1` | get the JSON representation of this resource | resources.json

I'll apply these changes to the table I made in the first comment. In the future, all JSON data related retrieving will be a GET request with `/data` at the end of it.

Wait - perhaps we should also standardize the query parameters. lol maybe I should make a separate issue for this. See #86

**Comment by aryam7** [comment]
Ok. This issue is very close to being done.
What do I have left to do?
- [ ] create a resource class for getting the JSON representation of a Resource (see issue #87)
- [ ] update the route that gets the JSON representation of a Resource so that it adheres to the standards in issue #86
- [ ] test that the store method works + write code for creating resource contents
- [ ] test that the update method works + handle null input

**Comment by aryam7** [comment]
Hey @GiselleSerate,
So I realized that I also need to give you the Resource Content ID in the JSON, so that you can properly make a request to edit the Resource Content (should you ever need to).
To summarize, this is what the JSON should look like now:
![image](https://user-images.githubusercontent.com/23412689/43497779-cccbedac-94f8-11e8-99aa-a32f752f797e.png)
Let me know if there's anything you want changed! Everything else should be the same as what you requested in issue #51.

**Comment by aryam7** [comment]
I realize that the image of the JSON I included above has a Resource Content that isn't a link even though the type of the Resource Content is "link". I should change the ResourceContent seeder so that it actually seeds links when it needs to. Also, it would be great if we validated that links were links (both while creating and updating).

**Comment by aryam7** [comment]
agh I also realized @GiselleSerate probably wanted a `use_name` in the returned JSON.

Here's the new (hopefully finalized) JSON:
```
{
  "meta": {
    "name": "Resource 2",
    "author_name": "Larissa Ziemann",
    "author_type": "teacher",
    "use_name": "Notes"
  },
  "contents": [
    {
      "id": 2,
      "name": "Resource Content 2",
      "type": "link",
      "content": "http://www.hessel.com/",
      "created": "Aug 1, 2018 3:08 AM",
      "updated": "Aug 1, 2018 3:08 AM"
    }
  ]
}
````

**Comment by ciennG** [comment]
I updated the functions in Server.js to use these new paths and such. Also fixed an inaccuracy in the code that assumed we were working in an older version of d3. A couple of notes: 

- Right now, server always retrieves resource information and returns resource information as JSON. I believe this was our intended behavior, but let me know if this needs to change for some implementations.
-  Functions that handle CRUD for resources all accept two optional callbacks (callback1 and callback2). If specified, callback1 will run on any returned error message if the request returns an error, or, if specified, callback2 will run on the result of the request if there was no error. If callback1 is not specified (or if the function is passed anything other than a function), the function throws any error it encounters. If callback2 is not specified, the result of the request will be returned unprocessed.

These functions aren't tested yet, but they can be found on Server.js in the Tree_Development branch for review.

**Comment by aryam7** [comment]
Ok. So new plan: give editing its own URI (`/resources/{id}/edit`) and let it use the same view as `resources.create`
